File: regions/system/library/go/eventstore/memory.go
package eventstore

import (
	"context"
	"sync"
)

// InMemoryEventStore はメモリ内イベントストアの実装。
type InMemoryEventStore struct {
	mu      sync.RWMutex
	streams map[string][]*EventEnvelope
}

// NewInMemoryEventStore は新しい InMemoryEventStore を生成する。
func NewInMemoryEventStore() *InMemoryEventStore {
	return &InMemoryEventStore{
		streams: make(map[string][]*EventEnvelope),
	}
}

func (s *InMemoryEventStore) Append(_ context.Context, streamID StreamId, events []*EventEnvelope, expectedVersion *uint64) (uint64, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	key := streamID.String()
	stream := s.streams[key]

	var currentVersion uint64
	if len(stream) > 0 {
		currentVersion = stream[len(stream)-1].Version
	}

	if expectedVersion != nil && *expectedVersion != currentVersion {
		return 0, NewVersionConflictError(*expectedVersion, currentVersion)
	}

	version := currentVersion
	for _, event := range events {
		version++
		// イベントのコピーを作成
		copy := *event
		copy.Version = version
		copy.StreamID = key
		stream = append(stream, &copy)
	}
	s.streams[key] = stream
	return version, nil
}

func (s *InMemoryEventStore) Load(_ context.Context, streamID StreamId) ([]*EventEnvelope, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	stream, ok := s.streams[streamID.String()]
	if !ok {
		return []*EventEnvelope{}, nil
	}
	result := make([]*EventEnvelope, len(stream))
	for i, e := range stream {
		copy := *e
		result[i] = &copy
	}
	return result, nil
}

func (s *InMemoryEventStore) LoadFrom(_ context.Context, streamID StreamId, fromVersion uint64) ([]*EventEnvelope, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	stream, ok := s.streams[streamID.String()]
	if !ok {
		return []*EventEnvelope{}, nil
	}
	var result []*EventEnvelope
	for _, e := range stream {
		if e.Version >= fromVersion {
			copy := *e
			result = append(result, &copy)
		}
	}
	if result == nil {
		return []*EventEnvelope{}, nil
	}
	return result, nil
}

func (s *InMemoryEventStore) Exists(_ context.Context, streamID StreamId) (bool, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	_, ok := s.streams[streamID.String()]
	return ok, nil
}

func (s *InMemoryEventStore) CurrentVersion(_ context.Context, streamID StreamId) (uint64, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	stream, ok := s.streams[streamID.String()]
	if !ok || len(stream) == 0 {
		return 0, nil
	}
	return stream[len(stream)-1].Version, nil
}

// InMemorySnapshotStore はメモリ内スナップショットストアの実装。
type InMemorySnapshotStore struct {
	mu        sync.RWMutex
	snapshots map[string]*Snapshot
}

// NewInMemorySnapshotStore は新しい InMemorySnapshotStore を生成する。
func NewInMemorySnapshotStore() *InMemorySnapshotStore {
	return &InMemorySnapshotStore{
		snapshots: make(map[string]*Snapshot),
	}
}

func (s *InMemorySnapshotStore) SaveSnapshot(_ context.Context, snapshot *Snapshot) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	copy := *snapshot
	s.snapshots[snapshot.StreamID] = &copy
	return nil
}

func (s *InMemorySnapshotStore) LoadSnapshot(_ context.Context, streamID StreamId) (*Snapshot, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	snap, ok := s.snapshots[streamID.String()]
	if !ok {
		return nil, nil
	}
	copy := *snap
	return &copy, nil
}

File: regions/system/library/rust/eventstore/src/memory.rs
use std::collections::HashMap;
use std::sync::Arc;

use async_trait::async_trait;
use tokio::sync::RwLock;

use crate::envelope::EventEnvelope;
use crate::error::EventStoreError;
use crate::snapshot::Snapshot;
use crate::store::EventStore;
use crate::stream::StreamId;

pub struct InMemoryEventStore {
    streams: Arc<RwLock<HashMap<String, Vec<EventEnvelope>>>>,
}

impl InMemoryEventStore {
    pub fn new() -> Self {
        Self {
            streams: Arc::new(RwLock::new(HashMap::new())),
        }
    }
}

impl Default for InMemoryEventStore {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl EventStore for InMemoryEventStore {
    async fn append(
        &self,
        stream_id: &StreamId,
        events: Vec<EventEnvelope>,
        expected_version: Option<u64>,
    ) -> Result<u64, EventStoreError> {
        let mut streams = self.streams.write().await;
        let key = stream_id.to_string();
        let stream = streams.entry(key).or_default();

        let current_version = stream.last().map(|e| e.version).unwrap_or(0);

        if let Some(expected) = expected_version {
            if expected != current_version {
                return Err(EventStoreError::VersionConflict {
                    expected,
                    actual: current_version,
                });
            }
        }

        let mut version = current_version;
        for mut event in events {
            version += 1;
            event.version = version;
            event.stream_id = stream_id.to_string();
            stream.push(event);
        }

        Ok(version)
    }

    async fn load(&self, stream_id: &StreamId) -> Result<Vec<EventEnvelope>, EventStoreError> {
        let streams = self.streams.read().await;
        let key = stream_id.to_string();
        Ok(streams.get(&key).cloned().unwrap_or_default())
    }

    async fn load_from(
        &self,
        stream_id: &StreamId,
        from_version: u64,
    ) -> Result<Vec<EventEnvelope>, EventStoreError> {
        let streams = self.streams.read().await;
        let key = stream_id.to_string();
        Ok(streams
            .get(&key)
            .map(|events| {
                events
                    .iter()
                    .filter(|e| e.version >= from_version)
                    .cloned()
                    .collect()
            })
            .unwrap_or_default())
    }

    async fn exists(&self, stream_id: &StreamId) -> Result<bool, EventStoreError> {
        let streams = self.streams.read().await;
        Ok(streams.contains_key(&stream_id.to_string()))
    }

    async fn current_version(&self, stream_id: &StreamId) -> Result<u64, EventStoreError> {
        let streams = self.streams.read().await;
        let key = stream_id.to_string();
        Ok(streams
            .get(&key)
            .and_then(|events| events.last())
            .map(|e| e.version)
            .unwrap_or(0))
    }
}

pub struct InMemorySnapshotStore {
    snapshots: Arc<RwLock<HashMap<String, Snapshot>>>,
}

impl InMemorySnapshotStore {
    pub fn new() -> Self {
        Self {
            snapshots: Arc::new(RwLock::new(HashMap::new())),
        }
    }
}

impl Default for InMemorySnapshotStore {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl crate::snapshot::SnapshotStore for InMemorySnapshotStore {
    async fn save_snapshot(&self, snapshot: Snapshot) -> Result<(), EventStoreError> {
        let mut snapshots = self.snapshots.write().await;
        snapshots.insert(snapshot.stream_id.clone(), snapshot);
        Ok(())
    }

    async fn load_snapshot(
        &self,
        stream_id: &StreamId,
    ) -> Result<Option<Snapshot>, EventStoreError> {
        let snapshots = self.snapshots.read().await;
        Ok(snapshots.get(&stream_id.to_string()).cloned())
    }
}
